django/http/response.py

```diff
@@ -443,29 +445,37 @@ def set_headers(self, filelike):
         Set some common response headers (Content-Length, Content-Type, and
         Content-Disposition) based on the `filelike` response content.
         """
-        encoding_map = {
-            'bzip2': 'application/x-bzip',
-            'gzip': 'application/gzip',
-            'xz': 'application/x-xz',
-        }
-        filename = getattr(filelike, 'name', None)
-        filename = filename if (isinstance(filename, str) and filename) else self.filename
-        if os.path.isabs(filename):
-            self['Content-Length'] = os.path.getsize(filelike.name)
+        filename = getattr(filelike, 'name', '')
+        filename = filename if isinstance(filename, str) else ''
+        if hasattr(filelike, 'seek') and (not hasattr(filelike, 'seekable') or filelike.seekable()):
+            if hasattr(filelike, 'tell'):
+                current_position = filelike.tell()
+                filelike.seek(0, io.SEEK_END)
+                self['Content-Length'] = filelike.tell()
+                filelike.seek(current_position)
+            else:
+                self['Content-Length'] = sum(iter(lambda: len(filelike.read(self.block_size)), 0))
+                filelike.seek(-self['Content-Length'], io.SEEK_END)
```

Comment:
Had to add some minor magic so it's possible to get back to the initial position when `tell` isn't available.